activesupport/lib/active_support/actionable_error.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+require "active_support/concern"
+
+module ActiveSupport
+  # Actionable errors let's you define actions to resolve an error.
+  #
+  # To make an error actionable, include the <tt>ActiveSupport::ActionableError</tt>
+  # module and invoke the +action+ class macro to define the action. An action
+  # needs a name and a block to execute.
+  module ActionableError
+    extend Concern
+
+    NonActionable = Class.new(StandardError)
+
+    NoActions = Hash.new do |_, name| # :nodoc:
+      raise NonActionable, "Cannot find action \"#{name}\""
+    end
+
+    included do
+      class_attribute :_actions, default: NoActions.dup
+    end
+
+    def self.actions(error) # :nodoc:
+      case error
+      when ActionableError, -> it { Class === it && it < ActionableError }
+        error._actions
+      else
+        NoActions
+      end
+    end
+
+    def self.dispatch(error, name) # :nodoc:
+      actions(error.is_a?(String) ? error.constantize : error)[name].call
+    end
+
+    module ClassMethods
+      # Defines an action that can resolve the error.
+      #
+      #   class PendingMigrationError < MigrationError
+      #     include ActiveSupport::ActionableError
+      #
+      #     action "Run pending migrations" do
+      #       ActiveRecord::Tasks::DatabaseTasks.migrate
+      #     end
+      #   end
+      def action(name, &block)
+        _actions[name] = block
+      end
```

Comment:
Should we eval the block in the instance context of the error? Is it conceivable that an action might want to reference data from the error? E.g. ActiveRecord::RecordInvalid's record attribute.

Reply:
Hm, that's not really possible because we'd need to reconstruct the error data in the next request…

Reply:
Yeah, that's why I didn't wanna complicate it even more.

Reply:
Yeah, good call, let's keep it the way it is now 👍